pr_ros_controllers/include/pr_ros_controllers/joint_mode_controller.h

```diff
@@ -0,0 +1,120 @@
+ *********************************************************************/
+
+/// \author Ethan Kroll Gordon
+
+#pragma once
+
+
+#include <vector>
+#include <string>
+
+#include <ros/node_handle.h>
+#include <realtime_tools/realtime_buffer.h>
+
+// actionlib
+#include <actionlib/server/action_server.h>
+
+// ROS messages
+#include <pr_control_msgs/JointModeCommandAction.h>
+
+// ros_controls
+#include <realtime_tools/realtime_server_goal_handle.h>
+#include <controller_interface/controller.h>
+#include <hardware_interface/joint_mode_interface.h>
+
+
+namespace pr_ros_controllers
+{
+
+/**
+ * \brief Joint mode controller for a set of joints.
+ *
+ * This class provides an actionlib interface the control mode of the provided joints.
+ *
+ *
+ * \section ROS interface
+ *
+ * \param joints Names of the joints to control.
+ * \param default String of mode to initialize all joints.
+ *                Possible values: BEGIN, POSITION, VELOCITY, EFFORT, NOMODE, OTHER, SWITCHING
+ *                All other values will be converted to ERROR
+ * \see hardware_interface::JointCommandModes
+ *
+ */
+
+class JointModeController: public controller_interface::Controller<hardware_interface::JointModeInterface>
+{
+public:
+  JointModeController() : controller_interface::Controller<hardware_interface::JointModeInterface>() {}
+  ~JointModeController() {}
+
+  bool init(hardware_interface::JointModeInterface* hw, ros::NodeHandle &n) override;
+
+  void starting(const ros::Time& /*time*/) { /* Do Nothing */}
+  void stopping(const ros::Time& /*time*/) { /* Do Nothing */}
+  void update(const ros::Time& /*time*/, const ros::Duration& /*period*/)
+  {
+    std::vector<hardware_interface::JointCommandModes> & commands = *commands_buffer_.readFromRT();
+    for(unsigned int i=0; i<n_joints_; i++)
+    {  joint_modes_[i].setMode(commands[i]); }
+  }
+
+protected:
+  typedef actionlib::ActionServer<pr_control_msgs::JointModeCommandAction>                    ActionServer;
+  typedef std::shared_ptr<ActionServer>                                                       ActionServerPtr;
+  typedef ActionServer::GoalHandle                                                            GoalHandle;
+
+  realtime_tools::RealtimeBuffer< std::vector<hardware_interface::JointCommandModes> > commands_buffer_;
+  unsigned int n_joints_;
+
+  std::vector< std::string >                         joint_names_;    ///< Controlled joint names.
+  std::vector< hardware_interface::JointModeHandle > joint_modes_;    ///< Handle to controlled joint.
+  std::string                                        name_;           ///< Controller name.
+
+  // ROS API
+  ros::NodeHandle    controller_nh_;
+  ActionServerPtr    action_server_;
+  ros::Timer         goal_handle_timer_;
+  ros::Duration      action_monitor_period_;
+
+  // Callback, should immediately send result
+  void goalCB(GoalHandle gh);
+
+  // General callbacks
+  void cancelCB(GoalHandle gh) { /* Do Nothing */ }
```

Comment:
I've been looking into what happens when a cancel request is received from the client.

It seems like ActionServer calls the GoalHandle's `setCancelRequested` function ([line 210 of action_server_imp.h](https://docs.ros.org/en/diamondback/api/actionlib/html/action__server__imp_8h_source.html#l00210)), and that function sets the Goal's status to either PREEMPTING or RECALLING (starting at [line 335 of server_goal_handle_imp.h](https://docs.ros.org/en/diamondback/api/actionlib/html/server__goal__handle__imp_8h_source.html#l00318)). So just FYI, this cancel callback won't actually do nothing, it will set the goal status to either PREEMPTING or RECALLING (even though in practice this code will neither try to preempt or recall the goal).

One option could be setting the goal status back to active in this CB. However, with that said, whether this is a problem or not depends on how you anticipate clients will use this action server. So if you think it won't be misleading to clients (either they won't cancel or they will except cancel to do nothing), then there is no need to change it.

Reply:
That's fair. Since the goal is immediately transfered to the DONE state in `goalCB`, this function should never be called anyway, but in case there are any weird issues, I explicitly added a `gh.setCanceled` to complete the lifecycle.